### ISSUE TITLE
Using id<MSALAccount> in completion blocks and MSALResult

### ISSUE THREAD

**Issue by br-earl-gaspard** [AzureAD/microsoft-authentication-library-for-objc#1299]
In my project I created a protocol called `Authenticator` and have `MSALPublicClientApplication` conform to it.

```
protocol Authenticator {
    func acquireToken(with parameters: MSALInteractiveTokenParameters, completionBlock: @escaping MSALCompletionBlock)
    func acquireTokenSilent(with parameters: MSALSilentTokenParameters, completionBlock: @escaping MSALCompletionBlock)
    func getCurrentAccount(with parameters: MSALParameters?, completionBlock: @escaping MSALCurrentAccountCompletionBlock)
    func signout(with account: MSALAccount, signoutParameters: MSALSignoutParameters, completionBlock signoutCompletionBlock: @escaping MSALSignoutCompletionBlock)
    func refreshAccessToken(with parameters: MSALSilentTokenParameters, completionBlock: @escaping MSALCompletionBlock)
}

extension MSALPublicClientApplication: Authenticator {}
```

This allows me to create a mock version for unit testing called `MockAuthenticator` that conforms to `Authenticator`. However, since `MSALCurrentAccountCompletionBlock` in `getCurrentAccount()` uses a concrete class for `MSALAccount` and `MSALAccount` has `NS_UNAVAILABLE` for its initializers, I am unable to create an `MSALAccount` or subclass for use in testing.

If the completion blocks in `MSALDefinitions.h` were setup to use the `MSALAccount` protocol then I believe testing `getCurrentAccount()` becomes easier.

```
typedef void (^MSALAccountsCompletionBlock)(NSArray<id<MSALAccount>> * _Nullable accounts, NSError * _Nullable error);
typedef void (^MSALCurrentAccountCompletionBlock)(id<MSALAccount> _Nullable account, id<MSALAccount> _Nullable previousAccount, NSError * _Nullable error);
```

Instead of:
```
typedef void (^MSALAccountsCompletionBlock)(NSArray<MSALAccount *> * _Nullable accounts, NSError * _Nullable error);
typedef void (^MSALCurrentAccountCompletionBlock)(MSALAccount * _Nullable account, MSALAccount * _Nullable previousAccount, NSError * _Nullable error);
```

Furthermore, `MSALResult` uses `MSALAccount` and not `id<MSALAccount>` which would effect testing calls using `MSALCompletionBlock`. Is that something can be changed too?

**Comment by hieunguyenmsft** [comment]
Thank you for the suggestion. I will look into that.

**Comment by hieunguyenmsft** [comment]
Hi @br-earl-gaspard, this is a public API and changing it will break other applications so we prefer not to do that. 
Thank you.

**Comment by br-earl-gaspard** [comment]
I respectfully disagree with the answer on two counts.
1. APIs are expected to introduce breaking changes from time to time. If they are not than they are not evolving and improving. [Semantic versioning](https://semver.org/) is a way to account for introducing incompatible API changes.
2. The changes proposed here with using a protocol instead of a concrete class do not seem like this would be a breaking change but more investigation may be needed.

**Comment by hieunguyenmsft** [comment]
Thanks for the suggestion. I think we're following that versioning model and changing public API will require a major version pump. Changing is not a problem here, we need to have a good reason to change, it will require other applications to update the source code.
The difference between protocol: MSALAccount and class MSALAccount is class MSALAccount is also conforming to other protocol NSCopying to support shallow copy object. Changing the return type to protocol type will not reflect that protocol and need client apps casting it back.
The justification of the benefit when changing to id<MSALAccount> aren't too many (testing is not totally blocked by it, I think with a subclass for mocking + Swizzle, the client app can still mock the account object returned).
Please let me know if the above makes sense to postpone the change right now (We can add it when we have another bigger update on public API). 
Thank you.

**Comment by br-earl-gaspard** [comment]
Thanks for the feedback @hieunguyenmsft. That helps clear up why the class is used instead of the protocol.
As far as testing goes, I'd prefer to stay away from Swizzling. Another option for the purpose of testing would be to remove the MSALAccount's NS_UNAVAILABLE setting. This would allow easy creation of subclasses but I understand NS_UNAVAILABLE was probably added for a reason.
